cpp/src_prims/sparse/distance/csr_spmv.cuh

```diff
@@ -0,0 +1,354 @@
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+#pragma once
+
+#include <raft/cudart_utils.h>
+#include <raft/sparse/cusparse_wrappers.h>
+#include <raft/cuda_utils.cuh>
+#include <raft/mr/device/allocator.hpp>
+#include <raft/mr/device/buffer.hpp>
+
+#include <sparse/distance/common.h>
+#include <sparse/utils.h>
+#include <sparse/csr.cuh>
+
+#include <limits.h>
+
+#include <nvfunctional>
+
+#include <cub/block/block_load.cuh>
+#include <cub/block/block_radix_sort.cuh>
+#include <cub/block/block_store.cuh>
+
+namespace raft {
+namespace sparse {
+namespace distance {
+
+/**
+ * Semiring which schedules each row of B in a different thread.
+ * @tparam value_idx
+ * @tparam value_t
+ * @tparam tpb
+ * @tparam buffer_size
+ * @tparam rows_per_block
+ */
+template <typename value_idx, typename value_t, int tpb, typename reduce_f,
+          typename accum_f>
+struct BlockSemiring {
+  __device__ inline BlockSemiring(int tid_, value_idx m_, value_idx n_,
+                                  value_idx *shared_cols_,
+                                  value_t *shared_vals_, value_idx *chunk_cols_,
+                                  value_t *chunk_vals_, value_idx *offsets_a_)
+    : tid(tid_),
+      p(p),
+      m(m_),
+      n(n_),
+      shared_cols(shared_cols_),
+      shared_vals(shared_vals_),
+      chunk_cols(chunk_cols_),
+      chunk_vals(chunk_vals_),
+      offsets_a(offsets_a_),
+      done(false),
+      shared_idx(0),
+      row_count(0),
+      cur_sum(0.0) {}
+
+  /**
+   * Load columns for a single row of A into shared memory
+   * @param row
+   * @param indptrA
+   * @param indicesA
+   * @param dataA
+   */
+  __device__ inline void load_a(value_idx row, value_idx *indptrA,
+                                value_idx *indicesA, value_t *dataA) {
+    if (tid == 0) {
+      offsets_a[0] = indptrA[row];
+      offsets_a[1] = indptrA[row + 1];
+    }
+    __syncthreads();
+
+    start_offset_a = offsets_a[0];
+    stop_offset_a = offsets_a[1];
+
+    shared_size = stop_offset_a - start_offset_a;
+
+    // Coalesce reads of row from matrix A into shared memory
+    for (int i = tid; i < shared_size; i += blockDim.x) {
+      shared_cols[i] = indicesA[start_offset_a + i];
+      shared_vals[i] = dataA[start_offset_a + i];
+    }
+
+    __syncthreads();
+
+    row_a = row;
+  }
+
+  /**
+   * Prepare index & offsets for looping through rows of B
+   * @param start_row
+   * @param indptrB
+   */
+  __device__ inline void load_b(value_idx start_row, value_idx *indptrB) {
+    done = false;
+    shared_idx = 0;
+    cur_sum = 0.0;
+
+    start_row_b = start_row;
+    stop_row_b = min(start_row_b + tpb, start_row_b + (n - start_row_b));
+
+    n_rows = (stop_row_b - start_row_b);
+
+    for (int i = tid; i < n_rows; i += tpb) {
```

Comment:
If I understand correctly, this executes only one iteration per thread. I feel it would be easier to follow if we had just an `if (tid < n_rows)`.